packages/block-library/src/post-excerpt/index.php

```diff
@@ -83,13 +83,16 @@ function register_block_core_post_excerpt() {
  * the excerpt length block setting has no effect.
  * Returns 100 because 100 is the max length in the setting.
  */
-if ( is_admin() ||
-	defined( 'REST_REQUEST' ) && REST_REQUEST ) {
-	add_filter(
-		'excerpt_length',
-		static function () {
-			return 100;
-		},
-		PHP_INT_MAX
-	);
-}
+add_filter(
+	'excerpt_length',
+	static function ( $value ) {
+		// This check needs to be inside the callback since the REST_REQUEST constant
+		// is not defined at the time add_filter() is called.
+		if ( ! ( defined( 'REST_REQUEST' ) && REST_REQUEST ) ) {
+			return $value;
+		}
+
+		return 100;
+	},
+	PHP_INT_MAX
+);
```

Comment:
The whole approach is still not great because the excerpt will be limited to 100 for any sort of REST request, even when not in the editor. Ideally this should only be added in `edit` context in the REST API. Plus. plugins cannot easily unhook it because it's an anonymous function added at PHP_INT_MAX.

Reply:
Also, unit tests would be great.

Reply:
> plugins cannot easily unhook it because it's an anonymous function added at PHP_INT_MAX

Yea, I know this is preexisting code but seems it would be good to remove the `PHP_INT_MAX` filter priority "nonsense" as it is a bad way to do things in WP, especially in core. 

Think that comes from some plugins trying to be "too clever" about adding their callback after everybody else. Unfortunately that doesn't work as any other plugin can use `PHP_INT_MAX` too and make sure it is loaded after the plugin that is trying to be clever :) 

Imho core should never use something like that, it's just a bad way to use filters. If something must run after all the filters have finished running, it can be hard-coded after the `apply_filters()` call. If not, a priority of 20 should be sufficient, no matter what.

Reply:
> The whole approach is still not great because the excerpt will be limited to 100 for any sort of REST request

Yea, this is the preexisting code. Enforces `100` for all excerpts that are used through REST. That seem to be needed in the editor only. Comes from https://github.com/WordPress/gutenberg/pull/44964. 

Seems this is sub-optimal. The (unresolved) question seems to be "Who has priority to set the excerpt length?" The user (in the editor) or a plugin that the user has enabled?

In any case the enforcing of 100 for excerpt length through REST should probably be removed by fixing/updating how that works in the editor.

Reply:
That's still quite a hacky workaround, why not just hook into the `rest_prepare_post` filter? Something like this (unlinted pseudo code):

```
function register_block_core_post_excerpt_length_filter( $value ) {
	return 100;
}

foreach (get_post_types() as $post_type ) {
	if ( post_type_supports( $post_type, 'excerpt' ) {
		add_filter(
			'rest_prepare_' . $post_type,
			static function( $response, $post, $request ) {
				$fields = $this->get_fields_for_response( $request );
				if ( 'edit' === $request['context'] && rest_is_field_included( 'excerpt', $fields ) ) {
					add_filter( 'excerpt_length', 'register_block_core_post_excerpt_length_filter', 20 );

					/** This filter is documented in wp-includes/post-template.php */
					$excerpt = apply_filters( 'get_the_excerpt', $post->post_excerpt, $post );
		
					/** This filter is documented in wp-includes/post-template.php */
					$excerpt = apply_filters( 'the_excerpt', $excerpt );

					$data = $response->get_data();
		
					$data['excerpt'] = array(
						'raw'       => $post->post_excerpt,
						'rendered'  => post_password_required( $post ) ? '' : $excerpt,
						'protected' => (bool) $post->post_password,
					);

					return rest_ensure_response( $data );
				}
				return $response;
			},
			10,
			3
		);
	}
}
```

Reply:
Thank you for your reply.
I'm afraid I fail to understand why you are calling the current approach hacky. Anyway, I tried to fix the bug without refactoring the way this callback is hooked into the filter. Refactoring it wasn't my goal.
Also, I could be wrong, but doesn't your proposed code apply the `get_the_excerpt` and `the_excerpt` filters twice? Aren't these filters already applied in `WP_REST_Posts_Controller::prepare_item_for_response()`?
https://github.com/WordPress/wordpress-develop/blob/ddb67a0099c9912c46a22374bcf90e5aac2f25e3/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L1874

Reply:
> I'm afraid I fail to understand why you are calling the current approach hacky.

Because it won't work for cases like REST API preloading or batch requests, where it could still have unintended side effects.

> but doesn't your proposed code apply the get_the_excerpt and the_excerpt filters twice

No, as we are getting the `$post->post_excerpt` again from scratch

Reply:
> Because it won't work for cases like REST API preloading or batch requests

You are right. Well spotted! Thank you. I've changed the code so now it should work with preloading and batch requests. I've tested it with preloading and it works, but it should also work for batch requests too.

> No, as we are getting the $post->post_excerpt again from scratch

I have two concerns about the proposed solution, which are just my personal opinion:

1. The solution completely rewrites the excerpt field, ignoring any controller-specific code. This **_might_** cause issues with custom third-party REST controllers that might have their own business logic for handling the excerpt field.

2. It applies filters twice (although I understand that we are getting `$post->post_excerpt` from scratch), which seems wrong to me from the architectural standpoint. Also, it **_might_** lead to performance issues.

Therefore, in this PR, I decided not to "overwrite" the excerpt field.

Reply:
Hooking into `rest_pre_dispatch` without checking whether it's actually a relevant request is way too broad again. You don't want to override the excerpt length for every single post ever returned in the REST API when in edit context.

At this point, it feels like we're riding in circles here. I recommend you to ask REST API maintainers in the REST API Slack channel for advice.


I suppose a more bullet proof way of filtering the excerpt length only when relevant is by introducing some sort of new query argument to force the limit to be applied. Something like `?force_excerpt_length=100` or so.

Reply:
On Friday, I updated the PR to make it more selective and only add the filter when there is an "excerpt" field in the response (borrowing this idea from your code snippet). I'm not able to come up with anything better that doesn't rewrite the excerpt field.

> I recommend you to ask REST API maintainers in the REST API Slack channel for advice.

Thank you, I've taken that advice into account. Honestly, I doubt there is another "hook" that would allow us not to overwrite the excerpt field and at the same time provide fine-grained control over the exact post type to which the filter should be applied.

_**But I will check that.**_

> At this point, it feels like we're riding in circles here.

Haha. I have the same feeling. I will leave implementing the solution with `force_excerpt_length` to another contributor, as I'm running out of steam.

I really appreciate your feedback, @swissspidy.